CRI: Properly restore IP information for userns sandboxes

dcantah opened the pull request:
Fixes: #10363

Due to when we setup networking for userns containers (AFTER the OCI runtime has ran), this was causing the underlying object for the sandbox to not have the IP information saved in the metadata extension. Because we're trying to move away from modifying the underlying container object for sandboxes (or even just keeping around a reference to it), I don't want to update the underlying container afterwards to include this. So, to remedy this we can update the sandboxes metadata after network setup runs for userns containers, and then our recovery logic just needs to be altered a bit.

Our recovery logic already checks containers with the sandbox label first, and then checks our shiny new sandbox store. Our IP information is stored in the sandbox store objects, so all we need to do is check if it exists, and if so and we already found a container for it (although without the IP information) then just update the sandbox in our in-memory store.

dcantah commented:
/retest

fuweid commented on internal/cri/server/sandbox_run.go:
```diff
@@ -358,6 +358,15 @@ func (c *criService) RunPodSandbox(ctx context.Context, r *runtime.RunPodSandbox
 		sandboxCreateNetworkTimer.UpdateSince(netStart)
 	}
 
+	if err := sandboxInfo.AddExtension(podsandbox.MetadataKey, &sandbox.Metadata); err != nil {
```
The L323 has updated sandbox for extension. do we still need this line?

dcantah replied:
It's before the IP gets assigned however (in setupPodNetwork). Perhaps we should just get rid of that metadata save. I'm hesitant to because of how fickle these saves seem to be..

fuweid replied:
Maybe this IP information has been broken since https://github.com/containerd/containerd/pull/7679.

I think we should add e2e test to cover it.

fuweid replied:
The L410 has added sandbox into store. The ip is from that store.

dcantah replied:
>Maybe this IP information has been broken since https://github.com/containerd/containerd/pull/7679.
I think so yea. That's a good idea, let me see how the other userns integration tests function. I added a recovery test to verify we still have ip information before but for regular containers.

dcantah replied:
>The L410 has added sandbox into store. The ip is from that store.

That's our in-memory store iirc. That's the reason before restart everything is fine, but after a restart that is not present anymore and we need to consult (and re-populate that in-memory store) with what's on disk.

dcantah commented:
@fuweid Sorry for delay, un-drafted this and added an integration test. Only some of the CI machines support userns in full so I added a log so we can grep for if the testcase is actually being hit. "restart_linux_test.go:36: adding userns pods for containerd restart test"

dcantah commented:
/retest

fuweid commented on internal/cri/server/restart.go:
```diff
@@ -100,12 +101,18 @@ func (c *criService) recover(ctx context.Context) error {
 	}
 	for _, sbx := range storedSandboxes {
-		if _, err := c.sandboxStore.Get(sbx.ID); err == nil {
+		sb, err := c.sandboxStore.Get(sbx.ID)
```
Suggestion: we should merge metadata-db.sandbox-store.metadata with metadata-db.container.metadata in function `podSandboxLoader.RecoverContainer`. This loop is designed for the sandboxes which are managed by remote sandbox plugin. I think we should not involve two sources in one loop.

In the `RecoverContainer` function, we should prefer to use metadata from metadata-db.sandbox. What do you think?

-----

@dmcgowan  @mxpv 

I found that the sandboxes bucket wasn't in `v1`. Should we add the version before namespace for sandbox?

https://github.com/containerd/containerd/blob/67a0efc589af5a1a092d3489cfe2d1e9ec01506d/core/metadata/buckets.go#L309-L315

dcantah replied:
I agree, but that seemed a big shift for this PR. This "use the underlying container for the given sandbox" vs "use the sandbox api as the source of truth to grab the sandboxes" is very confusing atm.

fuweid replied:
> use the underlying container for the given sandbox

It's to recover sandbox created by previous release actually. 

> but that seemed a big shift for this PR

OK. What about adding a condition that if `IP` is empty, we should use metadata from sandbox bucket 😂

dcantah replied:
I can give that a try!

dcantah commented on internal/cri/server/sandbox_run.go:
```diff
@@ -358,6 +350,15 @@ func (c *criService) RunPodSandbox(ctx context.Context, r *runtime.RunPodSandbox
 		sandboxCreateNetworkTimer.UpdateSince(netStart)
 	}
 
+	if err := sandboxInfo.AddExtension(podsandbox.MetadataKey, &sandbox.Metadata); err != nil {
```
Works for me. When I'm by my computer I'll see if there's any gotchas

AkihiroSuda commented:
Removing from v2.0 milestone as staled
